### ISSUE TITLE
Cannot protect default namespace and handle error in custom namespace client the same time

### ISSUE THREAD

**Issue by mitsos1os** [socketio/socket.io#2291]
I am developing a project with socket.io communication. The basic structure is that I have set up a custom namespace which requires authentication. I provide authentication through JWT (Json Web Tokens).

My problem is the following. Authentication is provided as namespace middleware that if it fails, it calls the `next` callback of socket middleware function with an `Error`. Everything works correctly if I implement the authentication inside the custom namespace. ex.:

``` javascript

    var io = require('socket.io')();
    io
        .of('/somePath')
        .use(function authenticate(socket,next){
            if (authenticate.(socket)){
                var err = new Error('UnauthorizedError');
                next(err);
            }
            else{
                next();
            }
        })
        .on('connection',function(){
            console.log('Socket successfully connected to custom namespace /somePath');
        });

    io.listen(3000);

    //Client Code
    var clientIO = require('socket.io-client');

    var client = clientIO('http://localhost:3000/somePath');
    client.on('connect',function(){
        console.log('Connected to host');
    });
    client.on('error',function(err){
        console.error(err);
    });
```

_The client receives the error normally from the namespace middleware._

**However** if I want to have more custom namespaces that all share the same authentication logic, I have to implement the logic for everyone of them separately. ex.:

``` javascript

    var io = require('socket.io')();
    io
        .of('/somePath')
        .use(function authenticate(socket,next){
            if (authenticate.(socket)){
                var err = new Error('UnauthorizedError');
                next(err);
            }
            else{
                next();
            }
        })
        .on('connection',function(){
            console.log('Socket successfully connected to custom namespce /somePath');
        });

    io
        .of('/someOtherPath')
        .use(function authenticate(socket,next){
            if (authenticate.(socket)){
                var err = new Error('UnauthorizedError');
                next(err);
            }
            else{
                next();
            }
        })
        .on('connection',function(){
            console.log('Socket successfully connected to custom namespce /someOtherPath');
        });

    io.listen(3000);

    //Client Code
    var clientIO = require('socket.io-client');

    var client = clientIO('http://localhost:3000/somePath');
    client.on('connect',function(){
        console.log('Connected to host');
    });
    client.on('error',function(err){
        console.error(err);
    });
```

This again works. The authentication problem here is that **this way, we have setup authentication only for the custom namespaces. The default namespace `/` still remains unprotected and clients can connect to it without any credentials.**

So, I thought, I should use the authentication mechanism as a middleware on the main namespace so that it will have to pass through that in order to proceed to the custom namespace connection. ex.:

``` javascript

    //Server Code
    var io = require('socket.io')();
    io.use(function(socket,next){
        var err = new Error('I want an error');
        next(err);
    })
    io.on('connection', function(socket){
        console.log('Connected on main namespace');
    });
    io.listen(3000);
    //Namespace
    io
        //First Path
        .of('somePath')
        .on('connection',function(){
            console.log('Socket successfully connected to custom namespce /somePath');
        });
        //Second path
        .of('someOtherPath')
        .on('connection',function(){
            console.log('Socket successfully connected to custom namespce /someOtherPath');
        })

    //Client Code
    var clientIO = require('socket.io-client');

    var client = clientIO('http://localhost:3000/somePath');
    client.on('connect',function(){
        console.log('Connected to host');
    });
    client.on('error',function(err){
        console.error(err);
    });
```

_Notice that the middleware is not set up on the main namespace and not on the other two custom namespaces_

This approach, resolves the connection to the main namespace problem without authentication, but creates an **other problem!**. From the way that socket.io connections work to custom namespaces, they first connect to the main namespace `/` and then connect to the custom provided namespace, (in this case `/somePath`, `/someOtherPath` . This way, the socket client (of custom namespace) confronts an error while first attempting to connect to the main namespace, and does not continue to attempt the connection to the custom namespace that we want it to connect to. **However**, on the client we have setup the error handler on the custom namespace, so the error is transmitted on the main namespace channel and not caught by the client error handler which is listening to the custom namespace, which in turn is never connected since main namespace connection failed.

**In conclusion, by setting the middleware on the main namespace to protect it from unauthorized use, the clients that try to connect to custom namespaces, do not receive the authentication error.**

Any ideas on how to circumvent that?

Thanks

**Comment by nkzawa** [comment]
:+1: I also feel socket of custom namespace should receive the `error` event in such a case.

**Comment by darrachequesne** [comment]
That issue was closed automatically. Please check if your issue is fixed with the latest release, and reopen if needed (with a [fiddle](https://github.com/darrachequesne/socket.io-fiddle) reproducing the issue if possible).
